src/frontend/src/optimizer/plan_node/plan_tree_node_v2.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use smallvec::SmallVec;
+
+pub trait PlanTreeNodeV2 {
+    type PlanRef;
+
+    fn inputs(&self) -> SmallVec<[Self::PlanRef; 2]>;
+    fn clone_with_inputs(&self, inputs: impl Iterator<Item = Self::PlanRef>) -> Self;
+}
+
+macro_rules! impl_plan_tree_node_v2_for_generic_unary_node {
+    ($node_type:ident, $input_feild:ident) => {
+        impl<P: Clone> crate::optimizer::plan_node::plan_tree_node_v2::PlanTreeNodeV2
+            for $node_type<P>
+        {
+            type PlanRef = P;
+
+            fn inputs(&self) -> smallvec::SmallVec<[Self::PlanRef; 2]> {
+                smallvec::smallvec![self.$input_feild.clone()]
+            }
+
+            fn clone_with_inputs(&self, mut inputs: impl Iterator<Item = Self::PlanRef>) -> Self {
+                let mut new = self.clone();
+                new.$input_feild = inputs.next().expect("expect exactly 1 input");
+                assert!(inputs.next().is_none(), "expect exactly 1 input");
+                new.clone()
+            }
+        }
+    };
+}
+
+macro_rules! impl_plan_tree_node_v2_for_generic_binary_node {
+    ($node_type:ident, $first_input_feild:ident, $second_input_feild:ident) => {
+        impl<P: Clone> crate::optimizer::plan_node::plan_tree_node_v2::PlanTreeNodeV2
+            for $node_type<P>
+        {
+            type PlanRef = P;
+
+            fn inputs(&self) -> smallvec::SmallVec<[Self::PlanRef; 2]> {
+                smallvec::smallvec![
+                    self.$first_input_feild.clone(),
+                    self.$second_input_feild.clone()
+                ]
+            }
+
+            fn clone_with_inputs(&self, mut inputs: impl Iterator<Item = Self::PlanRef>) -> Self {
```

Comment:
Do we really need the clone method?

Reply:
I think it is used for the PlanRef. Considering if you want to modify a subtree without downcast every planRef, you must give a method to do that..